File: scala/src/test/scala/io/github/configur8/ConfigurationTemplateTest.scala
package io.github.configur8

import io.github.configur8.Property._
import org.scalatest.{FunSpec, Matchers}

class ConfigurationTemplateTest extends FunSpec with Matchers {

  private val intProperty = integer("anInteger")

  describe("Configuration Template") {
    it("reification blows up when required value is not supplied") {
      intercept[Misconfiguration] {
        ConfigurationTemplate().requiring(intProperty).reify()
      }
    }

    it("reification successful when all values supplied") {
      ConfigurationTemplate().requiring(intProperty).withProp(intProperty, 1).reify()
    }

    it("requiring props is immutable") {
      val original = ConfigurationTemplate().requiring(intProperty)
      val updated = original.withProp(intProperty, 1)

      updated.reify().valueOf(intProperty) shouldBe 1
      intercept[Misconfiguration] {
        original.reify()
      }
    }

    it("with props is immutable") {
      val original = ConfigurationTemplate().withProp(intProperty, 1)
      val updated = original.withProp(intProperty, 2)

      original.reify().valueOf(intProperty) shouldBe 1
      updated.reify().valueOf(intProperty) shouldBe 2
    }
  }
}

File: java/src/test/java/io/github/configur8/ConfigurationTemplateTest.java
package io.github.configur8;

import static io.github.configur8.ConfigurationTemplate.configurationTemplate;
import static org.hamcrest.Matchers.containsString;
import static org.hamcrest.core.Is.is;
import static org.junit.Assert.assertThat;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;

public class ConfigurationTemplateTest {

    public static final Property<String> FOO = Property.string("FOO");

    @Rule
    public ExpectedException thrown = ExpectedException.none();

    @Test
    public void requiringPropsIsImmutable() {
        ConfigurationTemplate originalConfig = configurationTemplate().requiring(FOO);
        ConfigurationTemplate updatedConfig = originalConfig.withProp(FOO, "bar");

        assertThat(updatedConfig.reify().valueOf(FOO), is("bar"));

        thrown.expectMessage(containsString("No value supplied for key 'FOO'"));
        thrown.expect(Misconfiguration.class);
        assertThat(originalConfig.reify().valueOf(FOO), is("foo"));
    }

    @Test
    public void overridingPropsIsImmutable() {
        ConfigurationTemplate originalConfig = configurationTemplate().withProp(FOO, "foo");
        ConfigurationTemplate updatedConfig = originalConfig.withProp(FOO, "bar");

        assertThat(originalConfig.reify().valueOf(FOO), is("foo"));
        assertThat(updatedConfig.reify().valueOf(FOO), is("bar"));
    }
}